sw/device/lib/crypto/drivers/otbn.c

```diff
@@ -194,3 +229,56 @@ otbn_error_t otbn_set_ctrl_software_errs_fatal(bool enable) {
 
   return kOtbnErrorOk;
 }
+
+/**
+ * Checks if the OTBN application's IMEM and DMEM address parameters are valid.
+ *
+ * This function checks the following properties:
+ * - IMEM and DMEM ranges must not be "backwards" in memory, with the end
+ * address coming before the start address.
+ * - The IMEM range must be non-empty.
+ *
+ * @param app the OTBN application to check
+ * @return `kHardenedBoolTrue` if checks pass, otherwise `kHardenedBoolFalse`.
+ */
+static hardened_bool_t check_app_address_ranges(const otbn_app_t *app) {
+  // IMEM must not be backwards or empty.
+  if (app->imem_end <= app->imem_start) {
+    return kHardenedBoolFalse;
+  }
+  HARDENED_CHECK_LT(app->imem_start, app->imem_end);
+
+  // DMEM data section must not be backwards.
+  if (app->dmem_data_end < app->dmem_data_start) {
+    return kHardenedBoolFalse;
+  }
+  HARDENED_CHECK_LE(app->dmem_data_start, app->dmem_data_end);
+
+  return kHardenedBoolTrue;
+}
+
+otbn_error_t otbn_load_app(const otbn_app_t app) {
+  if (!check_app_address_ranges(&app)) {
+    return kOtbnErrorInvalidArgument;
+  }
+
+  // Ensure OTBN is idle.
+  OTBN_RETURN_IF_ERROR(otbn_assert_idle());
+
+  const size_t imem_num_words = app.imem_end - app.imem_start;
+  const size_t data_num_words = app.dmem_data_end - app.dmem_data_start;
+
+  OTBN_RETURN_IF_ERROR(otbn_imem_sec_wipe());
+  // IMEM always starts at zero.
+  otbn_addr_t imem_start_addr = 0;
+  OTBN_RETURN_IF_ERROR(
+      otbn_imem_write(imem_num_words, app.imem_start, imem_start_addr));
+
+  OTBN_RETURN_IF_ERROR(otbn_dmem_sec_wipe());
```

Comment:
I realize this is a refactoring PR and therefore this question is out of scope, but is there a reason the IMEM write was performed before the DMEM wipe? I'm assuming that IMEM is instruction memory and DMEM is data memory. It seems like a successful instruction write followed by a failed data wipe would leave OTBN in kindof a dangerous state.

Reply:
Good point, thanks for the detailed review @jrvanwhy. Are you concerned about potentially exposing stale data in DMEM? That clearing should be done by the previous client of OTBN. The goal of the data wipe here is to make sure that the initial values of the DMEM are not easily predictable, making FI attacks on the dmem_write operation below harder to control and reproduce. PLMK if you had something else in mind.

Reply:
Discussed this with @jadephilipoom offline and we decided to move the dmem_sec_wipe after imem_sec_wipe.

Reply:
> Are you concerned about potentially exposing stale data in DMEM?

Yes, that was my concern.

Obviously you two have much better context on this code than I do so I have no further input.